Update auth-related functions and models

thomascube opened the pull request:
Refactored auth-related functions as per the latest JMAP spec updates

The two commits bring the client behavior in sync with the changes in the JMAP spec announced in the [JMAP forum](https://groups.google.com/forum/#!topic/jmap-discuss/mgXsDqzPR6Y) and requested in issue #44.

thomascube committed:
Adapt data strcutures and models used during authentication after changes to the JMAP spec

The authentication procedure defined in the JMAP spec changed recently. Issue #44.

The 200 response requesting more authorization data has been changed as follows:

- The `methods` property is now an array of `AuthMethod` objects
- The `continuationToken` property was renamed to `loginId`

thomascube committed:
Use the X-JMAP authentication scheme to construct the Authorization headers

This makes the Authorization headers sent by the JMAP client compliant with
RFC 7235 and the most recent version of the JMAP protocol specification.

Plus added reference to issue #44 in Changelog entries.

ddolcimascolo commented:
Hi @thomascube, thx for helping us catching up with the spec :)

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -650,7 +650,7 @@ export default class Client {
       Accept: 'application/json; charset=UTF-8',
       'Content-Type': 'application/json; charset=UTF-8',
-      Authorization: this.authToken
+      Authorization: 'X-JMAP ' + this.authToken
```
Humm... This will break other authentication mechanisms. The user is still free to do authentication without the library and use oauth for instance. We should not enforce X-JMAP here. For instance in OpenPaas Linagora is using "Bearer" because we use a custom-made JWT authentication done outisde this library.

However, as we should use X-JMAP when JMAP authentication is used, you can enfore "X-JMAP" in the header in the `withAuthAccess` method that implicitely sets the authentication token. The method is almost always certainly used right after the authentication is successful...

WDYT?

thomascube replied:
Yes, I already expected this concern :-)

And yes, you're right, the `X-JMAP` scheme is only required "If the user authenticated via the JMAP auth protocol". I agree to your suggestion to therefore only send that scheme when initialized using the `withAuthAccess` method. So besides the `authToken` we should also store the `authScheme` as a property of the client instance and use that for composing the Authorization header. To make it complete, I suggest to extend the `withAuthenticationToken` method with an additional parameter stating the `authScheme` assigned to the given token.

I'll update the pull request if you agree.

ddolcimascolo replied:
Agreed

thomascube committed:
Send Authroization header with X-JMAP scheme only when authenticated via the JMAP auth protocol

The `X-JMAP` auth scheme is implicitly added when using `Client.withAuthAccess()` while
the scheme can be specified as additional argument with `Client.withAuthenticationToken()`.

If no scheme was specified, the Authorization header only contains the provided token value.

This is an addition to the changes requsted in #44.

thomascube committed:
Store all payload properties in AuthMethod model

Use Object.assign() to copy all properties from the server response
to the model instance. babel-plugin-object-assign will make sure this
is properly translates to ES5.

ddolcimascolo commented on lib/models/AuthMethod.js:
```diff
@@ -0,0 +1,20 @@
+    Utils.assertRequiredParameterIsPresent(payload, 'payload');
+    Utils.assertRequiredParameterHasType(payload.type, 'type', 'string');
+
+    Object.assign(this, payload);
```
Nice :)

ddolcimascolo commented:
:+1:

dbenchi commented:
:+1:
